### ISSUE TITLE
Modal tabbing and shift+tabbing behavior

### ISSUE THREAD

**Issue by andrewnding** [twbs/bootstrap#25277]
Not sure if this is intended behavior, but the outermost div with the modal class has a tabindex of -1, which seems to me that it should get initial focus when the modal opens, but should not be reachable again by tabbing. However you can tab to it, which makes it seem like the focus highlighting has disappeared.

Another related possible issue is that when shift+tabbing, nothing happens if you try to shift tab past the first focusable element (so that the focus is on the outer div with tabindex -1). I would expect the focus to go from the first focusable element to the last focusable element, much like how if you tab forward, it jumps from the last focusable element to the first focusable element.

Modal demo for reference: https://getbootstrap.com/docs/4.0/components/modal/#live-demo

**Comment by patrickhlauke** [comment]
`tabindex="-1"` should not be re-focusable by TAB/SHIFT+TAB. i suspect it's the JS that currently sets focus back to it (the JS that is meant to maintain focus inside the modal), as normally negative tabindex only indicates something as being programmatically focusable, and not in the regular focus cycle.

and yes, while not mandatory, it would probably be good if the focus handling "wrapped" the focus cycle, as per https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal (so that once focus is on last item, next TAB sets it back to the first item, and vice-versa).

**Comment by andrewnding** [comment]
Ok that's what I suspected for the `tabindex="-1"` behavior. So a followup question I would have is should the `tabindex="-1"` div get focused at all, or would it be preferable to have it just be skipped over entirely? To meet 508 standards for my project, I can't have the `tabindex="-1"` div ever get focused since there's no visual indicator for it when it is focused, and I imagine that other users might have the same problem. If this is a direction that would make sense to head in, I'd be glad to take a stab at an implementation that would not cause the focus to disappear, as well as backward-wrapped tabbing.

https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_focus_discernable_predictable

**Comment by patrickhlauke** [comment]
it is valid and desireable to have focus set to the dialog itself when it first opens - see https://github.com/w3c/aria-practices/issues/442. generally, even for wcag/508 compliant dialogs, that initial focus does not fail visible focus requirements (once the JS issue is addressed that re-cycles into the `tabindex="-1"` modal itself.

**Comment by andrewnding** [comment]
I see that makes sense. Yeah I would be ok with having the focus initially set to the containing div with `tabindex="-1"`, as long as that `tabindex="-1"` div does not get refocused while tabbing through. I can try working on the following issues then:
- Prevent the outer tabindex="-1" div from getting focused while tabbing through the modal (aside from it being initially focused)
- Allow backwards tabbing to wrap around when reaching the first focusable element

**Comment by b-creative** [comment]
> * Prevent the outer tabindex="-1" div from getting focused while tabbing through the modal (aside from it being initially focused)
> * Allow backwards tabbing to wrap around when reaching the first focusable element

How do I do this?

**Comment by patrickhlauke** [comment]
closing this issue here - tracking the problem in the meta-issue https://github.com/twbs/bootstrap/issues/28481
